quantum/pointing_device/pointing_device_modes.c

```diff
@@ -0,0 +1,640 @@
+            return PD_RIGHT;
+        } else {
+            return PD_LEFT;
+        }
+    } else {
+        if (pointing_modes[current_device].y > 0) {
+            return PD_UP;
+        } else {
+            return PD_DOWN;
+        }
+    }
+}
+
+/**
+ * @brief update dependent parameters of pointing_mode
+ *
+ * Will update the direction and divisor values based on mode id and h, and v axis values
+ *
+ */
+void pointing_mode_update(void) {
+    current_direction = get_pointing_mode_direction();
+    current_divisor   = get_pointing_mode_divisor();
+}
+
+/**
+ * @brief Access current pointing mode direction
+ *
+ * @return uint8_t direction
+ */
+uint8_t current_pointing_mode_direction(void) {
+    return current_direction;
+}
+
+/**
+ * @brief Access current pointing mode divisor
+ *
+ * @return uint8_t divisor
+ */
+uint8_t current_pointing_mode_divisor(void) {
+    return current_divisor;
+}
+
+/**
+ * @brief Tap keycodes based on mode axis values
+ *
+ * Will translate internal h & v axes into keycode taps.
+ * Uses input ordered DOWN, UP, LEFT, RIGHT and will output 1 keytap/divisor
+ * and will update the h and v values.
+ *
+ * NOTE: favours staying on axis and weakly favours the horizontal over the vertical axis
+ *
+ * @params kc_left[in]  uint16_t keycode for negative h
+ * @params kc_down[in]  uint16_t keycode for negative v
+ * @params kc_up[in]    uint16_t keycode for positive v
+ * @params kc_right[in] uint16_t keycode for positive h
+ */
+void pointing_tap_codes(uint16_t kc_left, uint16_t kc_down, uint16_t kc_up, uint16_t kc_right) {
+    uint16_t kc_direction = 0;
+    int16_t  count        = 0;
+
+    switch (current_pointing_mode_direction()) {
+        case PD_DOWN:
+            kc_direction = kc_down;
+            break;
+
+        case PD_UP:
+            kc_direction = kc_up;
+            break;
+
+        case PD_LEFT:
+            kc_direction = kc_left;
+            break;
+
+        case PD_RIGHT:
+            kc_direction = kc_right;
+            break;
+    }
+
+    switch (current_pointing_mode_direction()) {
+        case PD_DOWN ... PD_UP:
+            count = divisor_divide16(pointing_modes[current_device].y);
+            if (!count) return; // exit if accumulated y is too low
+            pointing_modes[current_device].y -= divisor_multiply16(count);
+            pointing_modes[current_device].x = 0;
+            break;
+        case PD_LEFT ... PD_RIGHT:
+            count = divisor_divide16(pointing_modes[current_device].x);
+            if (!count) return; // exit if accumulated x is too low
+            pointing_modes[current_device].x -= divisor_multiply16(count);
+            pointing_modes[current_device].y = 0;
+            break;
+    }
+    // skip if KC_TRNS or KC_NO (but allow for axes update above)
+    if (kc_direction < 2) return;
+
+    // tap codes
+    uint8_t taps = abs(count);
+    do {
+        tap_code16_delay(kc_direction, POINTING_TAP_DELAY);
```

Comment:
Honestly, I'd rather see something like how encoder map works.  Eg, with action_exec, as that supports *all* keycodes. 
https://github.com/qmk/qmk_firmware/blob/50ab2305557a5ac47f79136055b227db52dc550d/quantum/encoder.c#L170-L182

https://github.com/qmk/qmk_firmware/blob/50ab2305557a5ac47f79136055b227db52dc550d/quantum/keyboard.h#L62-L64

https://github.com/qmk/qmk_firmware/blob/50ab2305557a5ac47f79136055b227db52dc550d/quantum/keyboard.h#L87-L91

But use a single row (eg 251), and just cycle through 1-4 for each of the modes (0,1,2,3 for one mode, 4,5,6,7 for the next, etc).

Reply:
Ah yes I remember now that this came up before and I intended to go down this route or try and do something similar to how combo's are handled I'll try and get some version of this working with `action_exec` sometime this week.

I remember stalling on this change before as I didn't want to change too much with core code but I may just have to bite the bullet on this as it really does make the feature more complete.

Thanks for the reminder and the other improvements!

Reply:
Almost got this working.
Just starting first round of debugging now, was a bit trickier than I initially understood but you basically laid out everything I needed (Thanks for that btw).
Felt compelled to dig through until I understood the flow myself and then realized basically all the files I needed to change you already linked...

Reply:
Only thing that is bugging me right now is that I would need to remove or hard code any of the built in key event modes or just leave the original tap_code16 thing as a separate but redundant system (which I don't like the idea of).